arroyo-worker/src/connectors/kinesis/source/mod.rs

```diff
@@ -0,0 +1,508 @@
+                continue;
+            }
+            let get_shard_iterator_call = shard.set_shard_iterator(get_shard_iterator_call.clone());
+            futures.push(shard.update_shard_iterator(get_shard_iterator_call));
+        }
+
+        loop {
+            select! {
+                result = futures.select_next_some() => {
+                    let shard_id = result.name;
+                    match self.handle_async_result_split(shard_id,
+                        result.result.map_err(|e| UserError::new("Fatal Kinesis error", e.to_string()))?, ctx).await? {
+                        Some(future) => {
+                            futures.push(future);
+                        },
+                        None => {}
+                    }
+                },
+                _ = sleep(Duration::from_secs(60)) => {
+                    if self.sync_shards(ctx).await.is_err() {
+                        warn!("failed to sync shards");
+                    }
+                }
+                control_message = ctx.control_rx.recv() => {
+                    match control_message {
+                        Some(ControlMessage::Checkpoint(c)) => {
+                            debug!("starting checkpointing {}", ctx.task_info.task_index);
+                            let mut s = ctx.state.get_global_keyed_state('k').await;
+                            for (shard_id, shard_state) in &self.shards {
+                                s.insert(shard_id.clone(), shard_state.clone()).await;
+                            }
+                            if self.checkpoint(c, ctx).await {
+                                return Ok(SourceFinishType::Immediate);
+                            }
+                        },
+                        Some(ControlMessage::Stop { mode }) => {
+                            info!("Stopping kinesis source: {:?}", mode);
+
+                            match mode {
+                                StopMode::Graceful => {
+                                    return Ok(SourceFinishType::Graceful);
+                                }
+                                StopMode::Immediate => {
+                                    return Ok(SourceFinishType::Immediate);
+                                }
+                            }
+                        }
+                        Some(ControlMessage::Commit { epoch: _ }) => {
+                            unreachable!("sources shouldn't receive commit messages");
+                        }
+                        None => {
+
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn process_records(
+        &mut self,
+        get_records_output: GetRecordsOutput,
+        ctx: &mut Context<(), T>,
+    ) -> Result<Option<String>, UserError> {
+        let records = get_records_output.records.unwrap_or_default();
+        for record in records {
+            let data = record.data.unwrap().into_inner();
```

Comment:
Do you know in what cases `record.data` and `record.approximate_arrival_timestamp` would be None?

Reply:
I was going to say they were both required and this is just a deficiency of the Kinesis client, but that's not entirely accurate. Data is required, while ApproximateArrivalTimestamp is not. Looking through the Wayback machine, it looks like the field was added at the end of 2023. It seems safe to assume it is always present. https://aws.amazon.com/about-aws/whats-new/2015/09/amazon-kinesis-server-side-timestamp/